src/View/Widget/DateTimeWidget.php

```diff
@@ -86,16 +99,35 @@ public function render(array $data, ContextInterface $context): string
             'val' => null,
             'type' => 'datetime-local',
             'escape' => true,
+            'step' => null,
             'timezone' => null,
             'templateVars' => [],
         ];
 
-        if ($data['type'] === 'datetime-local' || $data['type'] === 'time') {
-            $data += ['step' => '1'];
+        if (!isset($data['step'])) {
+            $step = null;
+            if (isset($this->defaultStep[$data['type']])) {
+                $step = $this->defaultStep[$data['type']];
+            }
+
+            if (isset($data['fieldName'])) {
+                $schemaType = $context->type($data['fieldName']);
```

Comment:
Having the widget use db schema type feels wrong to me. 

Perhaps all this introspection should be done in `FormHelper` itself since it holds the db type to input type mapping. `FormHelper::_magicOptions()` sets `step` for `number` inputs depending on whether the db type is `decimal` or `float`. ` So it can do the same for datetime inputs too.

Reply:
Makes you wonder what the point of the $context parameter is!

Reply:
Yup, there's definitely some work needed with regards to context introspection and setting of input/widget attributes.

But that's beyond the scope of this PR and needs to be addressed separately.

With regards to your changes for consistency please just add required changes to FormHelper::_magicOptions().

Reply:
The `context` objects aim to abstract the different potential sources that FormHelper has to work with. For example ORM object, elasticsearch entities, or arrays.

I don't think using the context inside the widget is incorrect. We've used `_magicOptions` to consolidate logic so we don't have duplicate logic between widgets or to generate data based on view context.

Reply:
@markstory Should we always include the fieldname in data array for all widgets then? Because without the fieldname the context is unusable. Or maybe add an additional argument to render() for fieldname.